src/frontend/src/planner/query.rs

```diff
@@ -50,7 +52,37 @@ impl Planner {
             let func_dep = plan.functional_dependency();
             order = func_dep.minimize_order_key(order, &[]);
 
-            let limit = limit.unwrap_or(LIMIT_ALL_COUNT);
+            let limit = limit.unwrap_or(ExprImpl::literal_bigint(LIMIT_ALL_COUNT as i64));
+            if !limit.is_const() {
+                return Err(ErrorCode::ExprError(
+                        "expects an integer or expression that can be evaluated to an integer after LIMIT"
+                    .into(),
+                )
+                .into());
+            }
+            let limit_cast_to_bigint = limit.cast_explicit(DataType::Int64).map_err(|_| {
+                RwError::from(ErrorCode::ExprError(
+                        "expects an integer or expression that can be evaluated to an integer after LIMIT"
+                    .into(),
+                ))
+            })?;
+            let limit = match limit_cast_to_bigint.fold_const() {
+                Ok(Some(datum)) => {
```

Comment:
* This _validation_ logic is better placed in binder, and `BoundQuery::limit` remains an `Option<u64>`. Planner is in charge of _transforming_ SQL into a plan graph.
* We can do cast first and then `try_fold_const`, rather than using `is_const` and `fold_const` separately.
* The allowed casts in PostgreSQL is `assign` rather than `explicit`. (`int` requires `implicit`, `decimal` requires `assign`, `jsonb` requires `explicit`. I tested in PostgreSQL and `LIMIT '1'::jsonb` is not allowed.)
* When the evaluated output is `None`, which means SQL NULL, PostgreSQL treats it as no limit. Let's add a test case for this.
* Test error cases: wrong type, not const, eval error, negative

Reply:
> Test error cases: wrong type, not const, eval error, negative

wrong type, negative can be constructed, I added them

but "eval error and not const" seems to be avoided by the parsing stage already

Reply:
used "cast_assign" and "try_fold_const" instead